### ISSUE TITLE
Irrelevant instance predicate `NonNull` for `List`, `Vec`, ...

### ISSUE THREAD

**Issue by jamesmckinna** [agda/agda-stdlib#2092]
Proposal: we already have in `Data.List.Base`
```agda
null : List A → Bool
null [] = true
null _ = false
```
so is it time to add (cf. #1538 etc.)
```agda
NonNull : List A → Set _
NonNull = T ∘ not ∘ null
```
etc. 
with eg
```agda
instance
  nonNull⇒nonZero[length] : {{ _ : NonNull xs }} → NonZero (length xs)
  nonNull⇒nonZero[length] {xs = _ ∷ _} = _
```
etc.
together with some structure proxying for the type class for having a `null` method and a `NonNull` predicate... etc.

**Comment by jamesmckinna** [comment]
Re: #1538 and the choice of a `record`-based definition to avoid explicit import of `Data.Unit`:
the only reason to explicitly import would be to have the name `tt` in scope...
... but AFAIU, Agda will happily do (forced) argument inference on any sub-term of type `⊤` without ever having to explicitly mention either `⊤` or `tt`, because of `eta` for records... so `_` ALWAYS suffices?

I think that's why I used vanilla `T` in #2000, and in the above fragments, because instances will likewise ALWAYS be given by `_`... so no `import`s necessary, and the error messages don't really seem to be better for having the `record` wrapper (or are they?). Is this understanding correct?

**Comment by jamesmckinna** [comment]
See [the `list-refinement` branch](https://github.com/jamesmckinna/agda-stdlib/tree/list-refinement) for an initial sketch. Not quite the panacea I'd hoped, even for `scanr`...

**Comment by jamesmckinna** [comment]
Six months on, and having closed #2260 already, I'll close this for lack of any positive take-up.
